src/envs/screws.py

```diff
@@ -0,0 +1,509 @@
+            if screw == goal_screw:
+                color = 'gold'
+            else:
+                color = 'grey'
+            rect = plt.Rectangle((screw_x, screw_y),
+                                 screw_width,
+                                 screw_height,
+                                 edgecolor=color,
+                                 facecolor=color,
+                                 label=screw.name)
+            ax.add_patch(rect)
+
+        img = utils.fig2data(fig, dpi=150)
+        plt.close()
+        return [img]
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return {
+            self._ScrewPickupable, self._AboveReceptacle, self._HoldingScrew,
+            self._ScrewInReceptacle
+        }
+
+    @property
+    def goal_predicates(self) -> Set[Predicate]:
+        return {self._ScrewInReceptacle}
+
+    @property
+    def types(self) -> Set[Type]:
+        return {self._screw_type, self._gripper_type, self._receptacle_type}
+
+    @property
+    def options(self) -> Set[ParameterizedOption]:
+        return {
+            self._MoveToScrew, self._MoveToReceptacle, self._MagnetizeGripper,
+            self._DemagnetizeGripper
+        }
+
+    @property
+    def action_space(self) -> Box:
+        # dimensions: [dx, dy, magnetized vs. unmagnetized].
+        lowers = np.array([
+            -(self.rz_x_ub - self.rz_x_lb), -(self.rz_y_ub - self.rz_y_lb), 0.0
+        ],
+                          dtype=np.float32)
+        uppers = np.array([(self.rz_x_ub - self.rz_x_lb),
+                           (self.rz_y_ub - self.rz_y_lb), 1.0],
+                          dtype=np.float32)
+        return Box(lowers, uppers)
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        # NOTE: currently, the agent can only pick things up or
+        # drop them iff its staying still. This isn't particularly
+        # realistic...
+        dx, dy, magnetization = action.arr
+        if dx != 0.0 or dy != 0.0:
+            return self._transition_move(state, dx, dy)
+
+        if magnetization > 0.5:
+            return self._transition_magnetize(state)
+        return self._transition_demagnetize(state)
+
+    def _get_tasks(self, num_tasks: int, possible_num_screws: List[int],
+                   rng: np.random.Generator) -> List[Task]:
+        tasks = []
+
+        # Sample center locations for each of the screw clusters.
+        screw_cluster_xcenters = []
+        for _ in range(self._num_screw_clusters):
+            screw_cluster_xcenters.append(
+                rng.uniform(self.rz_x_lb,
+                            self.rz_x_ub - self._receptacle_width))
+
+        for _ in range(num_tasks):
+            screw_name_to_pos: Dict[str, Tuple[float, float]] = {}
+            num_screws = rng.choice(possible_num_screws)
+            for si in range(num_screws):
+                existing_xys = set(screw_name_to_pos.values())
```

Comment:
maybe just keep track of the existing xs in this loop, and worry about the ys and the names after all of the xs have been determined?

Reply:
i like the current structure; i think doing the y's separately looks clunky, and this way, we could easily extend to put things in different y positions as well in the future if necessary!